pythonwhat/probe.py

```diff
@@ -0,0 +1,218 @@
+import pprint as pp
+import itertools
+import inspect
+from functools import partial
+from collections import OrderedDict
+
+TEST_NAMES = [
+    "test_mc",
+    "test_or",
+    "test_with",
+    "test_list_comp",
+    "test_dict_comp",
+    "test_generator_exp",
+    "test_import",
+    "test_object",
+    "test_correct",
+    "test_if_else",
+    "test_if_exp",
+    "test_for_loop",
+    "test_function",
+    "test_print",
+    "test_function_v2",
+    "test_operator",
+    "test_try_except",
+    "test_data_frame",
+    "test_dictionary",
+    "test_while_loop",
+    "test_student_typed",
+    "test_object_accessed",
+    "test_output_contains",
+    "test_lambda_function",
+    "test_expression_result",
+    "test_expression_output",
+    "test_function_definition",
+    "test_object_after_expression"
+]
+
+SUB_TESTS = {
```

Comment:
An easy way would be to assume that functions (or functions w/o arguments) and Nodes are sub_tests, but there are also cases where dictionaries and lists of functions are valid sub-tests.

Alternatively, I thought about modifying the test_* functions like...

```{python}
from sub_test import SubTestMarker

def with(STUFF..., body=SubTestMarker, state=None):
    # CODE...
```

Where SubTestMarker marks that `body` is a sub-test argument. Currently the default arguments for subtests are None, so this is like creating a sort of NoneSubtest type.